src/palindrom-dom.js

```diff
@@ -91,15 +93,51 @@ var PalindromDOM = (function() {
   PalindromDOM.prototype = Object.create(Palindrom.prototype);
 
   /**
+   * DISABLED FOR NOW: we don't know when rendering actually finishes.
+   * It's left here for the hope of having synchronous rendering at some point in the future.
+   * ====
+   * we need to scroll asynchronously, because we need the document rendered to search for the anchored element
+   * and even though onReceive + applyPatch are sync, Polymer is not, it renders async-ly
+  PalindromNetworkChannel.prototype.scrollToAnchorOrTopAsync = function(link) {
+    this.scrollAsyncTimeout && clearTimeout(this.scrollAsyncTimeout);
+    if (window && window.document) {
+      var anchorIndex;
+      var anchor;
+      // does the URL have an anchor
+      if (link && (anchorIndex = link.indexOf('#')) > -1) {
+        anchor = link.substr(anchorIndex);
+      }
+      if (!anchor) {
+        window && window.scrollTo(0, 0);
+      } else {
+        // if somehow someone manages to navigate twice in a 100ms,
+        // we don't scroll for their first navigation, i.e de-bouncing 
+        
+        this.scrollAsyncTimeout = setTimeout(() => {
+          // does that anchor exist in the page?
+          const anchorTarget = document.querySelector(anchor); // look for #element-id
+          if (anchorTarget) {
+            anchorTarget.scrollIntoView();
+          } else {
+            window.scrollTo(0, 0);
+          }
+        }, 100);
+      }
+    }
+  };
+  */
+  /**
    * Push a new URL to the browser address bar and send a patch request (empty or including queued local patches)
    * so that the URL handlers can be executed on the remote
    * @param url
    */
   PalindromDOM.prototype.morphUrl = function(url) {
     history.pushState(null, null, url);
-    this.network.changeState(url);
+    this.network.changeCurrentUrl(url, onReceiveFinished);
   };
-
+  function onReceiveFinished() {
+    window && window.scrollTo(0, 0);
```

Comment:
Since the above code doesn't need anything from the network, it doesn't make sense to wait for the response. 

We could scroll to the top immediately, which results in better perceived performance.